File: imposter_fact_game.js
const FACT_SETS = {
  space: [
    "A day on Venus is longer than an entire Venusian year because the planet rotates incredibly slowly.",
    "If you could fly a plane straight up, it would only take a little over an hour to reach outer space.",
    "Neutron stars are so dense that a teaspoon of material from one would weigh billions of tonnes.",
    "Footprints left on the moon could last for millions of years because the moon has no atmosphere or weather.",
  ],
  animals: [
    "Octopuses have three hearts and blue blood that relies on copper instead of iron to carry oxygen.",
    "Cows form best friends and experience measurable stress when they are separated from them.",
    "Capybaras are so chill that in Japan some hot springs let them bathe with human visitors every winter.",
    "Male seahorses go through labor and give birth to hundreds of babies at once.",
  ],
  food: [
    "Honey never spoils; archaeologists have found jars in ancient tombs that are still perfectly edible.",
    "Ketchup was sold as medicine in the 1830s and was marketed as a cure for indigestion.",
    "Bananas are berries, botanically speaking, while strawberries are not.",
    "Worcestershire sauce is fermented with anchovies that dissolve almost completely during the process.",
  ],
  history: [
    "Napoleon was once attacked by a horde of rabbits during a planned hunting expedition gone wrong.",
    "In 1976, Los Angeles tried to paint its famous Hollywood sign yellow and red for the bicentennial celebrations.",
    "Oxford University is older than the Aztec Empire by at least two centuries.",
    "The shortest war in history ended after about 38 minutes between Britain and Zanzibar in 1896.",
  ],
  tech: [
    "The very first webcam pointed at a coffee pot at Cambridge so researchers could check if it was empty without walking over.",
    "USB was intentionally designed so you can try to plug it in up to three times before it actually fits.",
    "NASA's Apollo guidance computers had less processing power than a modern pocket calculator.",
    "The first alarm clock could only ring at 4 a.m.; the inventor was a man who had to wake up early for work.",
  ],
  tongue_twisters: [
    "Recite ‘Unique New York’ five times fast without stumbling.",
    "Say ‘red leather, yellow leather’ six times with perfect clarity.",
    "Deliver ‘She sells seashells by the seashore’ while keeping eye contact with the group.",
    "Three times quickly, say ‘Irish wristwatch’ without laughing at yourself.",
  ],
  try_not_to_laugh: [
    "Tell the group your most ridiculous childhood snack combination with a perfectly straight face.",
    "Read this groaner without breaking: “I used to be a baker, but I couldn’t make enough dough.”",
    "Stare down the player to your left and say, “Serious scientists seriously study silly string,” without cracking up.",
    "Share the silliest animal fact you know, but act like it’s a Nobel Prize discovery.",
  ],
};

function shuffle(array) {
  for (let i = array.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}

function sampleFacts(facts, count) {
  if (count <= 0) {
    return [];
  }

  if (facts.length >= count) {
    const candidates = shuffle([...facts]);
    return candidates.slice(0, count);
  }

  const pool = shuffle([...facts]);
  while (pool.length < count) {
    pool.push(facts[Math.floor(Math.random() * facts.length)]);
  }
  return pool.slice(0, count);
}

function titleCase(topic) {
  return topic
    .split(/[_\s]+/)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(" ");
}

function createRound(numPlayers, topicChoice) {
  const topics = Object.keys(FACT_SETS);
  const topic = topicChoice === "random"
    ? topics[Math.floor(Math.random() * topics.length)]
    : topicChoice;

  if (!FACT_SETS[topic]) {
    throw new Error("Unknown topic");
  }

  const imposterIndex = Math.floor(Math.random() * numPlayers);
  const facts = sampleFacts(FACT_SETS[topic], Math.max(0, numPlayers - 1));

  const assignments = [];
  let factIndex = 0;
  for (let i = 0; i < numPlayers; i += 1) {
    if (i === imposterIndex) {
      assignments.push({ isImposter: true });
    } else {
      assignments.push({ isImposter: false, prompt: facts[factIndex] });
      factIndex += 1;
    }
  }

  return { topic, imposterIndex, assignments };
}

function setupTopicOptions() {
  const topicSelect = document.getElementById("topic-choice");
  const fragment = document.createDocumentFragment();
  Object.keys(FACT_SETS)
    .sort()
    .forEach((topic) => {
      const option = document.createElement("option");
      option.value = topic;
      option.textContent = titleCase(topic);
      fragment.appendChild(option);
    });
  topicSelect.appendChild(fragment);
}

function main() {
  const form = document.getElementById("setup-form");
  const playerInput = document.getElementById("player-count");
  const topicSelect = document.getElementById("topic-choice");
  const namesInput = document.getElementById("player-names");
  const roundArea = document.getElementById("round-area");
  const topicPill = document.getElementById("topic-pill");
  const messageText = document.getElementById("message-text");
  const revealButton = document.getElementById("reveal-button");
  const nextButton = document.getElementById("next-button");
  const resetButton = document.getElementById("reset-button");
  const progressNote = document.getElementById("progress-note");

  let currentRound = null;
  let currentIndex = 0;
  let currentNames = [];
  let revealActive = false;
  function getDisplayName(index) {
    if (currentNames[index]) {
      return currentNames[index];
    }
    return `Player ${index + 1}`;
  }

  function showPromptForCurrentPlayer() {
    const name = getDisplayName(currentIndex);
    messageText.textContent = `Pass the device to ${name}. When they're ready, have them press and hold “Reveal”.`;
    revealButton.textContent = `Press & hold to reveal for ${name}`;
    revealButton.hidden = false;
    nextButton.hidden = true;
    revealButton.classList.remove("active");
    revealActive = false;
    progressNote.textContent = `${currentIndex} / ${currentRound.assignments.length} players briefed`;
  }

  form.addEventListener("submit", (event) => {
    event.preventDefault();
    const rawNames = namesInput.value
      .split(/[\n,]/)
      .map((entry) => entry.trim())
      .filter(Boolean);
    const numPlayers = rawNames.length > 0
      ? rawNames.length
      : Number.parseInt(playerInput.value, 10);
    if (!Number.isFinite(numPlayers) || numPlayers < 3) {
      alert("The game needs at least three players.");
      playerInput.focus();
      return;
    }

    playerInput.value = String(numPlayers);

    currentRound = createRound(numPlayers, topicSelect.value);
    currentIndex = 0;
    currentNames = rawNames.length > 0
      ? rawNames
      : Array.from({ length: numPlayers }, (_, idx) => `Player ${idx + 1}`);

    topicPill.textContent = `Secret topic: ${titleCase(currentRound.topic)}`;
    topicPill.hidden = false;
    roundArea.classList.add("active");
    resetButton.hidden = false;
    revealButton.hidden = false;
    nextButton.hidden = true;
    progressNote.textContent = `0 / ${currentRound.assignments.length} players briefed`;

    showPromptForCurrentPlayer();
  });
  function revealForCurrentPlayer() {
    if (!currentRound || revealActive) {
      return;
    }

    revealActive = true;
    revealButton.classList.add("active");
    const assignment = currentRound.assignments[currentIndex];
    const name = getDisplayName(currentIndex);
    const topicLabel = titleCase(currentRound.topic);
    if (assignment.isImposter) {
      messageText.textContent = `${name}, you're the imposter! Listen closely to everyone else's stories and craft a convincing bluff.`;
    } else {
      messageText.textContent = `${name}, share this ${topicLabel} prompt when it's your turn: ${assignment.prompt}`;
    }
    nextButton.hidden = true;
  }

  function hideForCurrentPlayer() {
    if (!currentRound || !revealActive) {
      return;
    }

    revealActive = false;
    revealButton.classList.remove("active");
    const name = getDisplayName(currentIndex);
    const totalPlayers = currentRound.assignments.length;
    const isLastPlayer = currentIndex === totalPlayers - 1;
    nextButton.textContent = isLastPlayer
      ? "Start discussion"
      : `Next player: ${getDisplayName(currentIndex + 1)}`;
    messageText.textContent = `All set, ${name}? Tap “${nextButton.textContent}” and pass the device.`;
    nextButton.hidden = false;
  }

  revealButton.addEventListener("pointerdown", (event) => {
    event.preventDefault();
    revealForCurrentPlayer();
  });

  ["pointerup", "pointerleave", "pointercancel"].forEach((eventName) => {
    revealButton.addEventListener(eventName, hideForCurrentPlayer);
  });

  revealButton.addEventListener("keydown", (event) => {
    if ((event.key === " " || event.key === "Enter") && !revealActive) {
      event.preventDefault();
      revealForCurrentPlayer();
    }
  });

  revealButton.addEventListener("keyup", (event) => {
    if ((event.key === " " || event.key === "Enter") && revealActive) {
      event.preventDefault();
      hideForCurrentPlayer();
    }
  });

  nextButton.addEventListener("click", () => {
    if (!currentRound) {
      return;
    }

    currentIndex += 1;
    const totalPlayers = currentRound.assignments.length;
    if (currentIndex >= totalPlayers) {
      messageText.textContent = "Everyone has their intel. Start the conversation and see if the imposter can stay hidden!";
      nextButton.hidden = true;
      revealButton.hidden = true;
      progressNote.textContent = `${totalPlayers} / ${totalPlayers} players briefed`;
      return;
    }

    progressNote.textContent = `${currentIndex} / ${totalPlayers} players briefed`;
    showPromptForCurrentPlayer();
  });

  resetButton.addEventListener("click", () => {
    currentRound = null;
    currentIndex = 0;
    currentNames = [];
    roundArea.classList.remove("active");
    topicPill.hidden = true;
    revealButton.hidden = true;
    nextButton.hidden = true;
    resetButton.hidden = true;
    progressNote.textContent = "";
    messageText.textContent = "Choose your player count, then press “Start new round” to begin.";
    revealButton.classList.remove("active");
    revealActive = false;
  });
}

setupTopicOptions();
window.addEventListener("DOMContentLoaded", main);

File: imposter_fact_game.py
"""Word association party game centred on real and improvised prompts.

This module implements a social deduction game inspired by "The Imposter".
Each round all players but one receive a humorous, real-world fact or playful
challenge that is linked to a shared topic.  The remaining player is the
imposter: they must invent a convincing story on the fly and bluff their way
through the group discussion.  The same generation logic powers the browser
version provided in ``imposter_fact_game.html``.

Running the module directly starts an interactive command-line experience that
walks players through a full round.  The :func:`generate_round` function drives
the assignment logic and is intentionally reusable for tests or for integrating
the game into a different interface.
"""

from __future__ import annotations

import argparse
import random
import textwrap
from dataclasses import dataclass
from typing import Iterable, List, Sequence


# A curated selection of topics and their related prompts.  Keep this in sync
# with the ``FACT_SETS`` constant in ``imposter_fact_game.js`` so the browser
# host and CLI host share the same material.
FACT_SETS: dict[str, tuple[str, ...]] = {
    "space": (
        "A day on Venus is longer than an entire Venusian year because the planet"
        " rotates incredibly slowly.",
        "If you could fly a plane straight up, it would only take a little over an"
        " hour to reach outer space.",
        "Neutron stars are so dense that a teaspoon of material from one would"
        " weigh billions of tonnes.",
        "Footprints left on the moon could last for millions of years because the"
        " moon has no atmosphere or weather.",
    ),
    "animals": (
        "Octopuses have three hearts and blue blood that relies on copper instead"
        " of iron to carry oxygen.",
        "Cows form best friends and experience measurable stress when they are"
        " separated from them.",
        "Capybaras are so chill that in Japan some hot springs let them bathe with"
        " human visitors every winter.",
        "Male seahorses go through labor and give birth to hundreds of babies at"
        " once.",
    ),
    "food": (
        "Honey never spoils; archaeologists have found jars in ancient tombs that"
        " are still perfectly edible.",
        "Ketchup was sold as medicine in the 1830s and was marketed as a cure for"
        " indigestion.",
        "Bananas are berries, botanically speaking, while strawberries are not.",
        "Worcestershire sauce is fermented with anchovies that dissolve almost"
        " completely during the process.",
    ),
    "history": (
        "Napoleon was once attacked by a horde of rabbits during a planned hunting"
        " expedition gone wrong.",
        "In 1976, Los Angeles tried to paint its famous Hollywood sign yellow and"
        " red for the bicentennial celebrations.",
        "Oxford University is older than the Aztec Empire by at least two centuries.",
        "The shortest war in history ended after about 38 minutes between Britain"
        " and Zanzibar in 1896.",
    ),
    "tech": (
        "The very first webcam pointed at a coffee pot at Cambridge so researchers"
        " could check if it was empty without walking over.",
        "USB was intentionally designed so you can try to plug it in up to three"
        " times before it actually fits.",
        "NASA's Apollo guidance computers had less processing power than a modern"
        " pocket calculator.",
        "The first alarm clock could only ring at 4 a.m.; the inventor was a man"
        " who had to wake up early for work.",
    ),
    "tongue_twisters": (
        "Recite 'Unique New York' five times fast without stumbling.",
        "Say 'red leather, yellow leather' six times with perfect clarity.",
        "Deliver 'She sells seashells by the seashore' while keeping eye contact"
        " with the group.",
        "Three times quickly, say 'Irish wristwatch' without laughing at"
        " yourself.",
    ),
    "try_not_to_laugh": (
        "Tell the group your most ridiculous childhood snack combination with a"
        " perfectly straight face.",
        "Read this groaner without breaking: 'I used to be a baker, but I couldn't"
        " make enough dough.'",
        "Stare down the player to your left and say, 'Serious scientists"
        " seriously study silly string,' without cracking up.",
        "Share the silliest animal fact you know, but act like it's a Nobel Prize"
        " discovery.",
    ),
}


@dataclass(frozen=True)
class PlayerAssignment:
    """Information given to a single player for a round."""

    is_imposter: bool
    prompt: str | None

    def display_message(self, topic: str) -> str:
        """Return the text shown to a player during the briefing phase."""

        if self.is_imposter:
            return textwrap.fill(
                "You drew the imposter card! Listen to everyone else's stories,"
                " then invent your own and try to blend in."
            )
        assert self.prompt is not None
        return textwrap.fill(
            f"Secret topic: {topic.title()}. Your prompt when the spotlight is on"
            f" you: {self.prompt}"
        )


@dataclass(frozen=True)
class RoundSetup:
    """Complete configuration for a single round of the game."""

    topic: str
    imposter_index: int
    assignments: tuple[PlayerAssignment, ...]

    def __post_init__(self) -> None:  # pragma: no cover - defensive programming
        if not 0 <= self.imposter_index < len(self.assignments):
            raise ValueError("Imposter index must point at a player assignment")


def _sample_facts(
    facts: Sequence[str],
    count: int,
    rng: random.Random,
) -> List[str]:
    """Return ``count`` facts, sampling without exhausting the source list."""

    if count <= 0:
        return []

    if len(facts) >= count:
        # Sample without replacement when we can provide unique facts.
        return rng.sample(list(facts), count)

    # When we need more facts than are available, provide every fact at least
    # once, then keep drawing with replacement for the remainder.  Shuffling the
    # base set up front keeps the distribution fair.
    pool = list(facts)
    rng.shuffle(pool)
    while len(pool) < count:
        pool.append(rng.choice(facts))
    return pool[:count]


def generate_round(
    num_players: int,
    *,
    rng: random.Random | None = None,
    fact_sets: dict[str, Sequence[str]] | None = None,
) -> RoundSetup:
    """Create the secret assignments for a single round of the game.

    Parameters
    ----------
    num_players:
        How many people are participating in the round.  Must be three or more
        so there is at least one imposter and two truth tellers.
    rng:
        Optional :class:`random.Random` instance used for reproducible tests.
    fact_sets:
        Optional custom fact pool.  Defaults to :data:`FACT_SETS`.
    """

    if num_players < 3:
        raise ValueError("The game needs at least three players")

    rng = rng or random.Random()
    available_facts = fact_sets or FACT_SETS
    topic, facts = rng.choice(tuple(available_facts.items()))

    imposter_index = rng.randrange(num_players)
    prompt_count = num_players - 1
    prompt_pool = _sample_facts(facts, prompt_count, rng)

    assignments: list[PlayerAssignment] = []
    prompt_iter = iter(prompt_pool)
    for idx in range(num_players):
        if idx == imposter_index:
            assignments.append(PlayerAssignment(is_imposter=True, prompt=None))
        else:
            assignments.append(
                PlayerAssignment(is_imposter=False, prompt=next(prompt_iter))
            )

    return RoundSetup(
        topic=topic,
        imposter_index=imposter_index,
        assignments=tuple(assignments),
    )


def run_cli(num_players: int, *, rng_seed: int | None = None) -> None:
    """Run the interactive command-line version of the game."""

    rng = random.Random(rng_seed)
    setup = generate_round(num_players, rng=rng)

    print("\nWelcome to the Fact Imposter!\n")
    print(
        textwrap.fill(
            "Pass the screen around. Each player will secretly read their prompt."
            " Keep a poker face and try to expose the imposter once discussion"
            " begins!\n"
        )
    )

    for player_number, assignment in enumerate(setup.assignments, start=1):
        input(f"Pass to Player {player_number} and press Enter when you're ready...")
        print()
        print(assignment.display_message(setup.topic))
        input("\nHide your prompt and press Enter to continue.")
        print("\n" + "-" * 60 + "\n")

    print(
        textwrap.fill(
            "Everyone has seen their prompt. Discuss, vote on who you think the"
            " imposter is, and then reveal the topic: "
            f"{setup.topic.title()}"
        )
    )


def parse_args(argv: Iterable[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Play a social deduction game where only one player must fake their"
            " prompt."
        )
    )
    parser.add_argument(
        "players",
        type=int,
        help="Number of players taking part in the round (minimum 3).",
    )
    parser.add_argument(
        "--seed",
        type=int,
        default=None,
        help="Optional random seed for predictable assignments.",
    )
    return parser.parse_args(list(argv) if argv is not None else None)


def main(argv: Iterable[str] | None = None) -> None:
    args = parse_args(argv)
    run_cli(args.players, rng_seed=args.seed)


if __name__ == "__main__":  # pragma: no cover - manual usage
    main()
